forge/src/trace/mod.rs

```diff
@@ -0,0 +1,371 @@
+    /// decoded parameters.
+    Decoded(String, Vec<String>),
+}
+
+impl Default for RawOrDecodedCall {
+    fn default() -> Self {
+        RawOrDecodedCall::Raw(Vec::new())
+    }
+}
+
+/// Raw or decoded return data.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum RawOrDecodedReturnData {
+    /// Raw return data
+    Raw(Vec<u8>),
+    /// Decoded return data
+    Decoded(String),
+}
+
+impl Default for RawOrDecodedReturnData {
+    fn default() -> Self {
+        RawOrDecodedReturnData::Raw(Vec::new())
+    }
+}
+
+impl fmt::Display for RawOrDecodedReturnData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            RawOrDecodedReturnData::Raw(bytes) => {
+                if bytes.is_empty() {
+                    write!(f, "()")
+                } else {
+                    write!(f, "0x{}", hex::encode(&bytes))
+                }
+            }
+            RawOrDecodedReturnData::Decoded(decoded) => write!(f, "{}", decoded.clone()),
+        }
+    }
+}
+
+/// A trace of a call.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct CallTrace {
+    /// The depth of the call
+    pub depth: usize,
+    /// Whether the call was successful
+    pub success: bool,
+    /// The name of the contract, if any.
+    ///
+    /// The format is `"<artifact>:<contract>"` for easy lookup in local contracts.
+    ///
+    /// This member is not used by the core call tracing functionality (decoding/displaying). The
+    /// intended use case is for other components that may want to process traces by specific
+    /// contracts (e.g. gas reports).
+    pub contract: Option<String>,
+    /// The label for the destination address, if any
+    pub label: Option<String>,
+    /// The destination address of the call
+    pub address: Address,
+    /// Whether the call was a contract creation or not
+    pub created: bool,
+    /// The value tranferred in the call
+    pub value: U256,
+    /// The calldata for the call, or the init code for contract creations
+    pub data: RawOrDecodedCall,
+    /// The return data of the call if this was not a contract creation, otherwise it is the
+    /// runtime bytecode of the created contract
+    pub output: RawOrDecodedReturnData,
+    /// The gas cost of the call
+    pub gas_cost: u64,
+}
+
+impl CallTrace {
+    /// Updates a trace given another trace
+    fn update(&mut self, new_trace: Self) {
+        self.success = new_trace.success;
+        self.address = new_trace.address;
+        self.created = new_trace.created;
+        self.value = new_trace.value;
+        self.data = new_trace.data;
+        self.output = new_trace.output;
+        self.address = new_trace.address;
+        self.gas_cost = new_trace.gas_cost;
+    }
+}
+
+impl fmt::Display for CallTrace {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.created {
+            write!(
+                f,
+                "{}new {}@{}",
```

Comment:
good opportunity to finally make the address debug, so that we see the entire thing?
```suggestion
                "{}new {}@{:?}",
```